node/test/util/stitch_util.js

```diff
@@ -0,0 +1,833 @@
+/*
+ * Copyright (c) 2016, Two Sigma Open Source
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * * Redistributions of source code must retain the above copyright notice,
+ *   this list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of git-meta nor the names of its
+ *   contributors may be used to endorse or promote products derived from
+ *   this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+"use strict";
+
+const assert  = require("chai").assert;
+const co      = require("co");
+const NodeGit = require("nodegit");
+
+const Commit              = require("../../lib/util/commit");
+const RepoAST             = require("../../lib/util/repo_ast");
+const RepoASTTestUtil     = require("../../lib/util/repo_ast_test_util");
+const RepoASTUtil         = require("../../lib/util/repo_ast_util");
+const StitchUtil          = require("../../lib/util/stitch_util");
+const SubmoduleConfigUtil = require("../../lib/util/submodule_config_util");
+const TreeUtil            = require("../../lib/util/tree_util");
+
+const FILEMODE            = NodeGit.TreeEntry.FILEMODE;
+
+function deSplitSha(sha) {
+    return sha.slice(0, 2) + sha.slice(3);
+}
+
+/**
+ *  - Replace `refs/stitched/converted/${splitSha(sha)}}` refs with their
+ *    equivalent logical mapping, e.g., if logical commit "1" maps to "aabb",
+ *    then we will remap `refs/stitched/converted/aa/bb` to
+ *    `refs/stitched/converted/1`.
+ */
+function refMapper(actual, mapping) {
```

Comment:
I do not think refs are necessarily the right thing to use for this (I'm putting this note here because this is where I happened to notice it, but I'm not just talking about the tests).  Updating packed refs is expensive -- the entire file must be rewritten.   Notes make more sense to me, since they are designed for this, and can be added in more like O(log) time.

Reply:
Yes, I like it. Will change.

Reply:
(actually, the packed-refs file only needs a rewrite when things are changed -- new refs are added loose -- but I still think notes make more sense)